src/Toml/Parser.hs

```diff
@@ -64,31 +64,53 @@ text_ = void . text
 -- TOML parser
 ----------------------------------------------------------------------------
 
--- Keys
-
-bareKeyP :: Parser Text
-bareKeyP = lexeme $ Text.pack <$> bareStrP
-  where
-    bareStrP :: Parser String
-    bareStrP = some $ alphaNumChar <|> char '_' <|> char '-'
-
-literalStringP :: Parser Text
-literalStringP = lexeme $ Text.pack <$> (char '\'' *> anyChar `manyTill` char '\'')
-
--- TODO: this parser is incorrect, it doesn't recognize all strings
-basicStringP :: Parser Text
-basicStringP = lexeme $ Text.pack <$> (char '"' *> anyChar `manyTill` char '"')
+-- Strings
 
 textP :: Parser Text
 textP = multilineBasicStringP <|> multilineLiteralStringP <|> literalStringP <|> basicStringP
 
--- adds " or ' to both sides
-quote :: Text -> Text -> Text
-quote q t = q <> t <> q
-
 nonControlCharP :: Parser Text
 nonControlCharP = Text.singleton <$> satisfy (not . isControl)
 
+escapeSequenceP :: Parser Text
+escapeSequenceP = char '\\' >> anyChar >>= \case
+                    'b'  -> pure "\b"
+                    't'  -> pure "\t"
+                    'n'  -> pure "\n"
+                    'f'  -> pure "\f"
+                    'r'  -> pure "\r"
+                    '"'  -> pure "\""
+                    '\\' -> pure "\\"
+                    'u'  -> hexUnicodeP 4
+                    'U'  -> hexUnicodeP 8
+                    c    -> fail $ "Invalid escape sequence: " <> "\\" <> [c]
+  where
+    hexUnicodeP :: Int -> Parser Text
+    hexUnicodeP n = count n hexDigitChar
+                    >>= \x -> case toUnicode $ hexToInt x of
+                          Just c  -> pure (Text.singleton c)
+                          Nothing -> fail $ "Invalid unicode character: " <> "\\" <> (if n == 4 then "u" else "U") <> x
+      where
+        hexToInt :: String -> Int
+        hexToInt xs = read $ "0x" ++ xs
+
+        toUnicode :: Int -> Maybe Char
+        toUnicode x
+          -- Ranges from "The Unicode Standard".
+          -- See definition D76 in Section 3.9, Unicode Encoding Forms.
+          | x >= 0      && x <= 0xD7FF   = Just (chr x)
+          | x >= 0xE000 && x <= 0x10FFFF = Just (chr x)
+          | otherwise                    = Nothing
+
+basicStringP :: Parser Text
+basicStringP = lexeme $ mconcat <$> (char '"' *> (escapeSequenceP <|> nonControlCharP) `manyTill` char '"')
+
+literalStringP :: Parser Text
+literalStringP = lexeme $ Text.pack <$> (char '\'' *> nonEolCharP `manyTill` char '\'')
+  where
+    nonEolCharP :: Parser Char
+    nonEolCharP = satisfy (\c -> c /= '\n' && c /= '\r')
```

Comment:
I don't like this line, I want this parser to fail only when [`eol`](https://hackage.haskell.org/package/megaparsec-6.5.0/docs/Text-Megaparsec-Char.html#v:eol) succeeds.

@ChShersh I couldn't think of an easier way to do that, do you have any ideas?

Reply:
@ghallak No, I don't know better way. You could ask on `megaparsec` issue tracker about this. Usually parser combinator libraries have inversions of parsers, like this:

```haskell
-- | Succeeds only when given parser fails.
notP :: Parser a -> Parser (Token s)
```